enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,460 @@
+---
+title: automatic-detection-of-disks-in-Local-Storage-Operator
+authors:
+  - "@ashishranjan738"
+  - "@rohantmp"
+  - "@jarrpa"
+reviewers:
+  - "@jsafrane"
+  - "@hekumar"
+  - "@chuffman"
+  - "@sapillai"
+  - "@leseb"
+  - "@travisn"
+  - "@rohantmp"
+  - "@jarrpa"
+approvers:
+  - "@jsafrane"
+  - "@hekumar"
+  - "@chuffman"
+creation-date: 2020-01-21
+last-updated: 2020-03-12
+status: implementable
+---
+
+# Automatic detection of Disks in Local-Storage-Operator 
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This is a proposal to enable automated discovery of local devices via the `VolumeDiscovery` CRD and sorting them into StorageClasses based on their characteristics via the `AutoDetectVolume` CRD. These CRDs will be implemented in the LocalStorageOperator(LSO)
+
+## Motivation
+
+The existing method of manually managing disks on nodes does not scale, and cannot be consumed by another operator.
+Choosing devices to be used should be handled on the platform so that multiple consumers of local storage don't conflict.
+To enable layering, we need information about potential devices, so that we have information about availability,
+and we can use that information to sort devices into storage classes based on their capabilities.
+
+### Goals
+
+- Discovery of potentially usable local disks on chosen nodes.
+- Automatic sorting of available disks into StorageClasses based on their characteristics.
+- Detect newly attached disks
+
+### Non-Goals
+
+- Inherently protect against confilict with provisioners that own local devices on the same nodes automatic detection is configured to run.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
```

Comment:
* This will match all newly attached AWS EBS PVs, just before kubelet formats them.
* This will match all local disks that already have PVs, but these PVs are not bound / used yet.

Reply:
done